File: PAS/src/main/java/pl/pas/services/UserService.java
package pl.pas.services;


import lombok.NoArgsConstructor;
import pl.pas.managers.UserManager;
import pl.pas.model.user.Administrator;
import pl.pas.model.user.Client;
import pl.pas.model.user.Employee;
import pl.pas.model.user.User;

import javax.enterprise.context.RequestScoped;
import javax.inject.Inject;
import javax.ws.rs.*;
import javax.ws.rs.core.MediaType;
import java.util.List;

@NoArgsConstructor
@RequestScoped
@Path("user")
public class UserService {

    @Inject
    private UserManager userManager;

    @GET
    @Path("getAllClients")
    @Produces({MediaType.APPLICATION_XML, MediaType.APPLICATION_JSON})
    public List<Client> getAllClients() {
        return userManager.getAllClients();
    }

    @GET
    @Path("getAllActiveClients")
    @Produces({MediaType.APPLICATION_XML, MediaType.APPLICATION_JSON})
    public List<Client> getAllActiveClients() {
        return userManager.getAllActiveClients();
    }

    @GET
    @Path("getAllEmployees")
    @Produces({MediaType.APPLICATION_XML, MediaType.APPLICATION_JSON})
    public List<Employee> getAllEmployees() {
        return userManager.getAllEmployees();
    }

    @GET
    @Path("getAllAdministrators")
    @Produces({MediaType.APPLICATION_XML, MediaType.APPLICATION_JSON})
    public List<Administrator> getAllAdministrators() {
        return userManager.getAllAdministrators();
    }

    @GET
    @Path("getUserById/{uuid}")
    @Produces({MediaType.APPLICATION_XML, MediaType.APPLICATION_JSON})
    public User getUser(@PathParam("uuid") long uuid) {
        return userManager.getUser(uuid);
    }

    @GET
    @Path("getUserByLogin/{login}")
    @Produces({MediaType.APPLICATION_XML, MediaType.APPLICATION_JSON})
    public User getUser(@PathParam("login") String login) {
        return userManager.getUser(login);
    }

    // TODO Post method should response something
    @POST
    @Path("addEmployee")
    @Consumes({MediaType.APPLICATION_XML, MediaType.APPLICATION_JSON})
    public void addEmployee(Employee employee) {
        userManager.addEmployee(employee.getLogin(), employee.getName(), employee.getLastName());
    }

    @POST
    @Path("addAdministrator")
    @Consumes({MediaType.APPLICATION_XML, MediaType.APPLICATION_JSON})
    public void addAdministrator(Administrator administrator) {
        userManager.addAdministrator(administrator.getLogin(), administrator.getName(), administrator.getLastName());
    }

    @POST
    @Path("addClient")
    @Consumes({MediaType.APPLICATION_XML, MediaType.APPLICATION_JSON})
    public void addClient(Client client) {
        userManager.addClient(client.getLogin(), client.getName(), client.getLastName(), client.getAge());
    }

    @PUT
    @Path("updateAdministrator/{id}")
    @Consumes({MediaType.APPLICATION_XML, MediaType.APPLICATION_JSON})
    public void updateEmployee(@PathParam("id") long id, Administrator administrator) {
        userManager.updateUser(userManager.getUser(id), administrator.getLogin(), administrator.getName(), administrator.getLastName());
    }

    @PUT
    @Path("updateEmployee/{id}")
    @Consumes({MediaType.APPLICATION_XML, MediaType.APPLICATION_JSON})
    public void updateEmployee(@PathParam("id") long id, Employee employee) {
        userManager.updateUser(userManager.getUser(id), employee.getLogin(), employee.getName(), employee.getLastName());
    }

    @PUT
    @Path("updateClient/{id}")
    @Consumes({MediaType.APPLICATION_XML, MediaType.APPLICATION_JSON})
    public void updateClient(@PathParam("id") long id, Client client) {
        userManager.updateClient(userManager.getUser(id), client.getLogin(), client.getName(), client.getLastName(), client.getAge());
    }
}

File: SPA/src/app/user.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Administrator } from './model/administrator';
import { Employee } from './model/employee';
import { User } from './model/user';

@Injectable({
  providedIn: 'root'
})
export class UserService {
  url = 'https://localhost:8181/PAS/api/user';
  http: HttpClient;

  constructor(http: HttpClient) {
    this.http = http;
  }

  getAllClients(): any {
    return this.http.get<any>(this.url + '/getAllClients');
  }

  getAllEmployees(): any {
    return this.http.get<any>(this.url + '/getAllEmployees');
  }

  getAllAdministrators(): any {
    return this.http.get<any>(this.url + '/getAllAdministrators');
  }

  getAllActiveClients(): any {
    return this.http.get<any>(this.url + '/getAllActiveClients');
  }

  getUserEtag(id: number | undefined): any {
    return this.http.get(this.url + '/getUserById/' + id, { observe: 'response' });
  }

  addAdministrator(newUser: Administrator): any {
    return this.http.post<any>(this.url + '/addAdministrator', newUser);
  }

  addEmployee(newUser: Employee): any {
    return this.http.post<any>(this.url + '/addEmployee', newUser);
  }

  addClient(newUser: User): any {
    return this.http.post<any>(this.url + '/addClient', newUser);
  }

  updateAdministrator(editUser: Administrator, etag: string): any {
    return this.http.put<any>(this.url + '/updateAdministrator/' + editUser.userId, editUser, { headers: { 'if-match': etag }});
  }

  updateClient(editUser: User, etag: string): any {
    return this.http.put<any>(this.url + '/updateClient/' + editUser.userId, editUser, { headers: { 'if-match': etag }});
  }

  updateEmployee(editUser: Employee, etag: string): any {
    return this.http.put<any>(this.url + '/updateEmployee/' + editUser.userId, editUser, { headers: { 'if-match': etag }});
  }
}
